### ISSUE TITLE
MontePython apparently "forgets" scf parameters

### ISSUE THREAD

**Issue by arsalanadil** [brinckmann/montepython_public#139]
Hi everyone,

A few days ago I ran into a seemingly absurd issue that I've since found a "dirty fix" for. I'll describe my fix below in case others also run into this issue or if people have suggestions. 

We are testing out a particular quintessence model and are using MontePython for parameter extraction. The issue is that after O(10^4) steps, MontePython returns the following error and exits out of the entire run:
`classy.CosmoSevereError:`
`Error in Class: Class did not read input parameter(s): attractor_ic_scf, scf_tuning_index, scf_parameters`
This is of course really weird, since CLASS most definitely read those parameters for the previous thousands of steps. 

The issue seems to be coming from the shooting algorithm in the input.c module of CLASS where, if MP wanders into a particularly "bad" part of parameter space, the shooting algorithm returns "nan" for Omega_scf. So I believe this is causing some confusion in Classy where, instead of returning a CosmoComputationError (i.e. accept the "bad" point but assign it zero likelihood), it returns a "CosmoSevereError" (which causes the run to abort entirely) since it thinks that CLASS did not read the input parameters. 

The workaround that I have implemented for now is in the Classy module where, if the scf_parameters were present in the input param file, the error that's returned is CosmoComputationError and MP moves on to the next point. However, perhaps a better solution would be to implement a "smarter" shooting algorithm so that this issue is caught in CLASS itself. 

Happy to hear other suggestions or to learn if other people have faced this issue.

Best wishes,
Arsalan

**Comment by brinckmann** [comment]
Hi Arsalan,

Thanks for the report! Did you specify a sampling range where the CLASS computation doesn't fail, or do these errors occur at random points in parameter space where you just get unlucky? As you may know, you specify this as the 2nd and 3rd entries in the list in the parameter file:
`data.parameters['h']         	= [0.703,  0.6, 0.8, 0.0065,1,   'cosmo`']
where in this example 0.6 is the lower prior and 0.8 is the upper prior. In this case, MontePython will only sample the parameter space between the two values.

If the latter (occurs randomly), then I think you should post this on the CLASS github page [https://github.com/lesgourg/class_public/issues](https://github.com/lesgourg/class_public/issues) as a fix would require changes on that side. 

E

Best,
Thejs

**Comment by arsalanadil** [comment]
Hi Thejs,

Thanks for the prompt response. 

Indeed, I did specify the bounds, but since the minimum of our particular potential is highly sensitive to the particular value of the varying parameter, it is difficult to know a-priori where the parameter space would be particularly bad (one could argue that that is why one would use MCMC in the first place!). Moreover, since there is some degeneracy in the parameters, there is almost always some pathological combination that MP will eventually wander into. In this sense, the issue is not "random". 

The only reason I chose to post it here rather than on the CLASS page is because, to CLASS's credit, it does return a "shooting failed" error. But it does so while returning "nan" for Omega_scf. And my (possibly faulty) understanding is that the latter triggers a CosmoSevereError in Classy. 

A related issue that I am facing is the following: the shooting parameter being used is the overall scale of the potential (for example the V_0 in V(\phi) = V_0 e^(\lambda \phi) ). CLASS will shoot and return different values for this parameter depending on other scf parameters passed to it (for example \lambda). Therefore, this parameter is _not_ being sampled over in MP, but I would still want to see the final posterior distribution for the successful values returned from the shooting. Is that just a matter of adding this as a 'derived' parameter in the input file (and possibly adding a corresponding line in Classy)? Basically the same way that h is for 100*\theta_s?  

Thanks a bunch!

Best,
Arsalan
